pkg/ccl/backupccl/manifest_handling.go

```diff
@@ -223,52 +223,89 @@ func decompressData(ctx context.Context, mem *mon.BoundAccount, descBytes []byte
 ) (BackupManifest, int64, error) {
-	// The backup manifest should only be found in the base directory,
-	// unlike the checkpoints which can be found in both the base and
-	// progress directory depending on the version.
-	var r ioctx.ReadCloserCtx
-	var err error
-	if filename != backupManifestName {
-		if r, err = readLatestCheckpointFile(ctx, exportStore, filename); err != nil {
+	checkpointFile, err := readLatestCheckpointFile(ctx, exportStore, filename)
+	if err != nil {
+		return BackupManifest{}, 0, err
+	}
+
+	// Look for a checksum, if one is not found it could be an older backup,
+	// but we want to continue anyway.
+	checksumFile, err := readLatestCheckpointFile(ctx, exportStore, filename+backupManifestChecksumSuffix)
+	if err != nil {
+		if !errors.Is(err, cloud.ErrFileDoesNotExist) {
 			return BackupManifest{}, 0, err
 		}
-	} else {
-		if r, err = exportStore.ReadFile(ctx, filename); err != nil {
+		// Pass checksumFile as nil to indicate it was not found.
+		return readManifest(ctx, mem, exportStore, encryption, checkpointFile, nil)
+	}
+
+	return readManifest(ctx, mem, exportStore, encryption, checkpointFile, checksumFile)
+}
+
+// readBackupManifest reads and unmarshals a BackupManifest from filename in
+// the provided export store.
+func readBackupManifest(
+	ctx context.Context,
+	mem *mon.BoundAccount,
+	exportStore cloud.ExternalStorage,
+	filename string,
+	encryption *jobspb.BackupEncryptionOptions,
+) (BackupManifest, int64, error) {
+	manifestFile, err := exportStore.ReadFile(ctx, filename)
+	if err != nil {
+		return BackupManifest{}, 0, err
+	}
+
+	// Look for a checksum, if one is not found it could be an older backup,
+	// but we want to continue anyway.
+	checksumFile, err := exportStore.ReadFile(ctx, filename+backupManifestChecksumSuffix)
+	if err != nil {
+		if !errors.Is(err, cloud.ErrFileDoesNotExist) {
 			return BackupManifest{}, 0, err
 		}
+		// Pass checksumFile as nil to indicate it was not found.
+		return readManifest(ctx, mem, exportStore, encryption, manifestFile, nil)
 	}
-	defer r.Close(ctx)
-	descBytes, err := mon.ReadAll(ctx, r, mem)
+
+	return readManifest(ctx, mem, exportStore, encryption, manifestFile, checksumFile)
+}
+
+// readManifest reads and unmarshals a BackupManifest from filename in
+// the provided export store. If the passed bound account is not nil, the bytes
+// read are reserved from it as it is read and then the approximate in-memory
+// size (the total decompressed serialized byte size) is reserved as well before
+// deserialization and returned so that callers can then shrink the bound acct
+// by that amount when they release the returned manifest.
+func readManifest(
+	ctx context.Context,
+	mem *mon.BoundAccount,
+	exportStore cloud.ExternalStorage,
+	encryption *jobspb.BackupEncryptionOptions,
+	manifestReader ioctx.ReadCloserCtx,
+	checksumReader ioctx.ReadCloserCtx,
+) (BackupManifest, int64, error) {
+	defer manifestReader.Close(ctx)
```

Comment:
lets move this defer to after where the reader is initialized. This method is just taking the reader as a param, it shouldn't be responsible for closing it.